### ISSUE TITLE
`truthy-bool` false positive when checking loop variable after iterating over empty iterable

### ISSUE THREAD

**Issue by DetachHead** [python/mypy#16997]
```py
# mypy: enable-error-code=truthy-bool

class Foo: ...

items = list[Foo]()
item = None
for item in items:
    ...
if item: # error: truthy-bool, but it's actually false at runtime
    print(1)
```
[playground](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict%2Ccheck-untyped-defs%2Cdisallow-any-decorated%2Cdisallow-any-expr%2Cdisallow-any-explicit%2Clocal-partial-types%2Cstrict-equality%2Cwarn-incomplete-stub%2Cwarn-redundant-casts%2Cwarn-return-any%2Cwarn-unreachable%2Cwarn-unused-configs%2Cwarn-unused-ignores&gist=9a3488dae833cfed96751c50c1af3e8a)

**Comment by KotlinIsland** [comment]
More generally, the subject of the `for` loop is incorrectly narrowed to the container type, when it should be a union of the two:
```py
items = list[int]()
item = None
for item in items: ...
reveal_type(item)  # int
```

**Comment by Hnasar** [comment]
Thanks for the report! I believe this is a duplicate of https://github.com/python/mypy/issues/16321
